pkg/controller/disruption/disruption.go

```diff
@@ -0,0 +1,486 @@
+	glog.Infof("deletePod %q -> PDB %q", pod.Name, pdb.Name)
+	dc.enqueuePdb(pdb)
+}
+
+func (dc *DisruptionController) enqueuePdb(pdb *policy.PodDisruptionBudget) {
+	key, err := controller.KeyFunc(pdb)
+	if err != nil {
+		glog.Errorf("Cound't get key for PodDisruptionBudget object %+v: %v", pdb, err)
+		return
+	}
+	dc.queue.Add(key)
+}
+
+func (dc *DisruptionController) getPdbForPod(pod *api.Pod) *policy.PodDisruptionBudget {
+	pdbs, err := dc.pdbLister.GetPodPodDisruptionBudgets(pod)
+	if err != nil {
+		glog.Infof("No PodDisruptionBudgets found for pod %v, PodDisruptionBudget controller will avoid syncing.", pod.Name)
+		return nil
+	}
+
+	// TODO(mml): handle n>1 matches
+	return &pdbs[0]
+}
+
+func (dc *DisruptionController) getPodsForPdb(pdb *policy.PodDisruptionBudget) ([]*api.Pod, error) {
+	sel, err := unversioned.LabelSelectorAsSelector(pdb.Spec.Selector)
+	if sel.Empty() {
+		return []*api.Pod{}, nil
+	}
+	if err != nil {
+		return []*api.Pod{}, err
+	}
+	podList, err := dc.podLister.Pods(pdb.Namespace).List(sel)
+	if err != nil {
+		return []*api.Pod{}, err
+	}
+	pods := []*api.Pod{}
+	for _, p := range podList.Items {
+		obj, err := api.Scheme.DeepCopy(p)
+		if err != nil {
+			return []*api.Pod{}, err
+		}
+		pod := obj.(api.Pod)
+		pods = append(pods, &pod)
+	}
+	return pods, nil
+}
+
+func (dc *DisruptionController) worker() {
+	for {
+		func() {
+			key, quit := dc.queue.Get()
+			if quit {
+				return
+			}
+			defer dc.queue.Done(key)
+			glog.Infof("Syncing DB %q", key.(string))
+			if err := dc.sync(key.(string)); err != nil {
+				glog.Errorf("Error syncing PodDisruptionBudget %v, requeuing: %v", key.(string), err)
+				// TODO(mml): In order to be safe in the face of a total inability to write state
+				// changes, we should write an expiration timestamp here and consumers
+				// of the PDB state (the /evict subresource handler) should check that
+				// any 'true' state is relatively fresh.
+
+				// TODO(mml): file an issue to that effect
+				dc.queue.Add(key)
+			}
+		}()
+	}
+}
+
+func (dc *DisruptionController) sync(key string) error {
+	startTime := time.Now()
+	defer func() {
+		glog.V(4).Infof("Finished syncing PodDisruptionBudget %q (%v)", key, time.Now().Sub(startTime))
+	}()
+
+	obj, exists, err := dc.pdbLister.Store.GetByKey(key)
+	if !exists {
+		return err
+	}
+	if err != nil {
+		glog.Errorf("Unable to retrieve PodDisruptionBudget %v from store: %v", key, err)
+		return err
+	}
+
+	pdb := obj.(*policy.PodDisruptionBudget)
+
+	if err := dc.trySync(pdb); err != nil {
+		return dc.failSafe(pdb)
+	}
+
+	return nil
+}
+
+func (dc *DisruptionController) trySync(pdb *policy.PodDisruptionBudget) error {
+	pods, err := dc.getPodsForPdb(pdb)
+	if err != nil {
+		return err
+	}
+
+	expectedCount, desiredHealthy, err := dc.getExpectedPodCount(pdb, pods)
+	if err != nil {
+		return err
+	}
+
+	currentHealthy := dc.countHealthyPods(pdb, pods)
+	err = dc.updatePdbSpec(pdb, currentHealthy, desiredHealthy, expectedCount)
```

Comment:
So is all of this, simply to update the state of the disruption budget annotation given the state of the system?  

Because it seems super heavy weight, you're watching a ton of stuff that other controllers are already watching.